### PR TITLE
Re-work tpp semantics and operations

### PR DESCRIPTION
**Pull request by chelini**
Ready for RV

### CODE REVIEW

File: lib/TPP/ConvertLinalgToTpp.cpp [hunk 1]
```diff
@@ -288,8 +286,11 @@ struct ConvertGenericOpToTpp : public OpRewritePattern<linalg::GenericOp> {
       return success();
     }
     if (tpp::utils::isTppAdd(linalgOp)) {
+      // A = A + B
```
Comment (rengolin) [inline lib/TPP/ConvertLinalgToTpp.cpp:289]: IIRC from the discussion yesterday, `A = A + B` requires `A.shape == B.shape`, but for `C = A + B`, we could have `A.rank < B.rank`, with automatic broadcast (if the lower rank dims match).

The definition above uses `SameTypeOperands`, so the latter case would be invalid, which is ok as a first approximation, but if/when we have that case, we'll probably need to have an explicit broadcast before the TPP operations, and match both ops to the auto-broadcast TPP op.

File: tpp-rt/XsmmRunnerUtils.cpp [hunk 2]
```diff
@@ -259,47 +262,6 @@ _mlir_ciface_xsmm_unary_scalar_invoke(const libxsmm_datatype dType,
   kernel(&param);
 }
 
-LIBXSMM_INLINE void matrix_copy_NC_to_NCNC(float *src, float *dst, int T, int N,
```
Comment (rengolin) [inline tpp-rt/XsmmRunnerUtils.cpp:262]: This looks like an unrelated change...

Comment (chelini) [reply tpp-rt/XsmmRunnerUtils.cpp:262]: Yes, house cleaning. I can do it separately if needed.

Comment (rengolin) [reply tpp-rt/XsmmRunnerUtils.cpp:262]: Easy NFC commit outside of PR.

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 3]
```diff
@@ -84,21 +100,14 @@ def Tpp_IdentityOp : Tpp_Op<"identity"> {
...
   let hasVerifier = 1;
 }
 
 //===----------------------------------------------------------------------===//
 // ReluOp
 //===----------------------------------------------------------------------===//
 
-def Tpp_ReluOp : Tpp_Op<"relu"> {
+def Tpp_ReluOp : Tpp_UnaryOp<"relu", [SameTypeOperands]> {
   let summary = "Applies a Rectified Linear Unit function in place.";
```
Comment (adam-smnk) [inline include/TPP/Dialect/Tpp/TppOps.td:111]: I think some verifier should check that input and output are the same destination for unary ops.
As an example, currently `tpp.relu` happily accepts two different arguments and further lowering just implicitly ignores the input.

Comment (adam-smnk) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: Slightly modified case from `tpp-to-loops` that will pass:
```// CHECK: func.func @relu_to_loops(
func.func @relu_to_loops(%arg0: memref<3x3xf32>, %arg1: memref<3x3xf32>) {
  // CHECK-DAG: %[[ub:.*]] = arith.constant 3 : index
  // CHECK-DAG: %[[lb:.*]] = arith.constant 0 : index
  // CHECK-DAG: %[[step:.*]] = arith.constant 1 : index
  // CHECK-DAG: %[[relu:.*]] = arith.constant 0.000000e+00 : f32
  // CHECK: scf.for %[[i:.*]] = %[[lb]] to %[[ub]] step %[[step]] {
  // CHECK:   scf.for %[[j:.*]] = %[[lb]] to %[[ub]] step %[[step]] {
  // CHECK:     %[[load:.*]] = memref.load %[[ARG0]][%[[i]], %[[j]]] : memref<3x3xf32>
  // CHECK:     %[[max:.*]] = arith.maxf %[[load]], %[[relu]] : f32
  // CHECK:     memref.store %[[max]], %[[ARG0]][%[[i]], %[[j]]] : memref<3x3xf32>
  tpp.relu ins(%arg1: memref<3x3xf32>) out(%arg0: memref<3x3xf32>)
  return
}
```

Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: ops that's wrong the load should be from %arg1 and the store to %arg0. thx

Comment (adam-smnk) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: So, it is no longer in place only op, right? If that's the case, I think the TableGen docs need slight updates too.

Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: yes, is not longer in-place but can be if you pass the same subview/memref as input and output.

Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: will update the .td doc.

Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:111]: The point is that you can still have in-place beh. if you pass ins and out the same memref. But the new spec is 1 input and 1 output.

**Review by adam-smnk — APPROVED** [review]
lgtm

**Comment by chelini** [comment]
@rengolin this is ready for review when you have time

**Review by rengolin — COMMENTED** [review]
Looking great, thanks! Some comments inline, but nothing serious.

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 4]
```diff
@@ -314,15 +314,53 @@ static bool allIndexingsAreProjectedPermutation(linalg::GenericOp genericOp) {
...
-    return false;
   if (linalgOp.hasTensorSemantics())
     return false;
-  if (!hasStaticShape(linalgOp))
+  return hasStaticShape(linalgOp);
+}
+
+// Return true if the operation is binary.
+bool isBinaryOp(linalg::GenericOp linalgOp) {
+  if ((linalgOp.getNumDpsInputs() == 1) && (linalgOp.getNumDpsInits() == 1)) {
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:328]: The uses are the one in the region of the generic. Just want to make sure we are using the input/output instead of other values, because the lianlg.generic is non-isolated from above and can capture outside IR.

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 5]
```diff
@@ -314,15 +314,53 @@ static bool allIndexingsAreProjectedPermutation(linalg::GenericOp genericOp) {
...
+    Value outArg = linalgOp.getRegionOutputArgs()[0];
+    SmallVector<Operation *> usersOut(outArg.getUsers().begin(),
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp]: used std::distance.

File: lib/TPP/ConvertTppToXsmm.cpp [hunk 6]
```diff
@@ -449,14 +457,15 @@ struct ConvertTppAddOp : public OpRewritePattern<tpp::AddOp> {
...
-
+    Type outputType = addOp.getOut().getType();
+    assert(outputType.isa<MemRefType>() && "expect a memref type");
     MemRefType outputMemRef = outputType.cast<MemRefType>();
-    int64_t m = outputMemRef.getShape()[0];
-    int64_t n = outputMemRef.getShape()[1];
+    assert((outputMemRef.getRank() == 1 || outputMemRef.getRank() == 2) &&
+           "expect memref with rank 1 or 2");
+    
+    int64_t m = (outputMemRef.getRank() == 2) ? outputMemRef.getShape()[0] : 1;
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/ConvertTppToXsmm.cpp:463]: I am planning to change this code once more when we increase tpp ops. If you don't have string feeling on this I can come back to it later.

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 7]
```diff
@@ -29,47 +29,63 @@ def TppBRGEMMVNNIMemrefInput : StaticMemRefRankOf<[AnyFloat], [4]>;
...
+  let hasVerifier = 1;
 }
 
 //===----------------------------------------------------------------------===//
 // IdentityOp
 //===----------------------------------------------------------------------===//
 
-def Tpp_IdentityOp : Tpp_Op<"identity"> {
+def Tpp_IdentityOp : Tpp_UnaryOp<"identity"> {
   let summary = "Copies input to output.";
```
Comment (nhasabni) [inline include/TPP/Dialect/Tpp/TppOps.td:89]: I think `tpp.identity` also supports broadcasting. If so, can you please update the description to mention about it?

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 8]
```diff
@@ -29,47 +29,63 @@ def TppBRGEMMVNNIMemrefInput : StaticMemRefRankOf<[AnyFloat], [4]>;
...
-    }];
+    Example:
 
-    let extraClassDeclaration = [{
-      Value getOutput() { return getRhs(); };
-    }];
+    ```mlir
+
+    tpp.add ins(%1 : memref<2x2xf32>, %2: memref<2x2xf32>) 
+            out(%2: memref<2x2xf32>)
```
[the remark this answers is not in the document]
Comment (nhasabni) [reply include/TPP/Dialect/Tpp/TppOps.td:76]: I see the code for mapping `linalg.generic` into `tpp.add`. Looks like we are supporting both in-place and out-of-place add. If so, can you please clarify it in the description and also add an example of out-of-place add?

File: lib/TPP/Dialect/Tpp/TppUtils.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:336]: Makes sense.

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 9]
```diff
@@ -314,15 +314,55 @@ static bool allIndexingsAreProjectedPermutation(linalg::GenericOp genericOp) {
...
+}
+
+static bool hasZeroUser(Value val) {
+  return std::distance(val.getUsers().begin(), val.getUsers().end()) == 0;
+}
+
+// Return true if the operation is binary.
+static bool isBinaryOp(linalg::GenericOp linalgOp) {
+  if ((linalgOp.getNumDpsInputs() == 1) && (linalgOp.getNumDpsInits() == 1)) {
+    Value outArg = linalgOp.getRegionOutputArgs()[0];
```
Comment (rengolin) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:337]: If the region (of a non-binary linalg region) has no output arguments, this will crash. You only check for DPS inputs in the if expression.

Perhaps it could be easier to just put a simple check outside of both IFs in the first line:
```
  if (linalgOp.getOutputs().size() < 1)
    return false;
  Value outArg = linalgOp.getRegionOutputArgs()[0];
  ...
```

File: test/Conversion/TppToLoops/tpp-to-loops.mlir [hunk 10]
```diff
@@ -1,149 +1,202 @@
...
-  // CHECK:   }
-  // CHECK: }
+  // CHECK:     %[[load1:.*]] = memref.load %[[ARG0]][%[[i]], %[[j]]] : memref<3x3xf32>
+  // CHECK:     %[[load2:.*]] = memref.load %[[ARG1]][%[[i]], %[[j]]] : memref<3x3xf32>
+  // CHECK:     %[[add:.*]] = arith.addf %[[load1]], %[[load2]] : f32
+  // CHECK:     memref.store %[[add]], %[[ARG1]][%[[i]], %[[j]]] : memref<3x3xf32>
+  tpp.add ins(%arg0: memref<3x3xf32>, %arg1: memref<3x3xf32>) out(%arg1: memref<3x3xf32>)
+  return
+}
+
```
Comment (nhasabni) [inline test/Conversion/TppToLoops/tpp-to-loops.mlir:72]: Can you also add a test for `add_to_loops` with out-of-place `add`? Looks like it is missing.

**Comment by nhasabni** [comment]
Some comments, otherwise LGTM.

File: include/TPP/Dialect/Tpp/TppOps.td (hunk 8 above)
[the remark this answers is not in the document]
Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:76]: Update doc.

File: lib/TPP/Dialect/Tpp/TppUtils.cpp (hunk 9 above)
Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:337]: I think, at least in the generic, getNumDpsInits is the number of output: https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L221

Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:337]: https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/include/mlir/Interfaces/DestinationStyleOpInterface.td#L78

Comment (adam-smnk) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:337]: Yeah, it looks like inits map to output - I found this example a bit clearer: https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp#L558

Although, I wouldn't mind some more obvious API for that.

Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:337]: Welcome to MLIR

File: lib/TPP/ConvertTppToXsmm.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/ConvertTppToXsmm.cpp:463]: Makes sense.

File: include/TPP/Dialect/Tpp/TppOps.td (hunk 7 above)
Comment (rengolin) [reply include/TPP/Dialect/Tpp/TppOps.td:100]: I'm not seeing this change in the current state.

Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:100]: Added doc here: https://github.com/plaidml/tpp-mlir/pull/258/commits/0e9e2107a4c6c6d0d4014bdd304c870668388ea9

Comment (rengolin) [reply include/TPP/Dialect/Tpp/TppOps.td:100]: Oh, I must have been looking at the wrong time. Github isn't always smart on email links.

**Review by rengolin — APPROVED** [review]
LGTM, thanks!
